File: backend/app/middleware/privacy.py
"""
Privacy Controls for Backend Performance Monitoring

Implements:
- PII scrubbing from logs, traces, and metrics
- Sampling controls
- Data retention policies
"""

import re
import hashlib
import logging
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Any, Pattern
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)


# ===== CONFIGURATION =====

@dataclass
class PrivacyConfig:
    """Privacy configuration for backend monitoring."""
    
    # Enable/disable all tracking
    enabled: bool = True
    
    # PII patterns to scrub
    pii_patterns: List[Pattern] = field(default_factory=list)
    
    # Fields to always redact in logs/traces
    sensitive_fields: List[str] = field(default_factory=list)
    
    # Headers to never log
    sensitive_headers: List[str] = field(default_factory=list)
    
    # Maximum payload size to log (bytes)
    max_payload_size: int = 1024
    
    # Anonymize user IDs in metrics
    anonymize_user_ids: bool = True
    
    # Hash salt for anonymization
    anonymization_salt: str = ""
    
    # Retention periods (hours)
    trace_retention_hours: int = 24
    metric_retention_hours: int = 168  # 7 days
    
    def __post_init__(self):
        if not self.pii_patterns:
            self.pii_patterns = DEFAULT_PII_PATTERNS
        if not self.sensitive_fields:
            self.sensitive_fields = DEFAULT_SENSITIVE_FIELDS
        if not self.sensitive_headers:
            self.sensitive_headers = DEFAULT_SENSITIVE_HEADERS


# Default patterns
DEFAULT_PII_PATTERNS: List[Pattern] = [
    # Email addresses
    re.compile(r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}', re.IGNORECASE),
    # Phone numbers
    re.compile(r'(\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}'),
    # SSN
    re.compile(r'\b\d{3}[-.\s]?\d{2}[-.\s]?\d{4}\b'),
    # Credit card numbers
    re.compile(r'\b\d{4}[-.\s]?\d{4}[-.\s]?\d{4}[-.\s]?\d{4}\b'),
    # JWT tokens
    re.compile(r'eyJ[a-zA-Z0-9_-]*\.eyJ[a-zA-Z0-9_-]*\.[a-zA-Z0-9_-]*'),
    # IP addresses (optional - might be needed for debugging)
    # re.compile(r'\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b'),
]

DEFAULT_SENSITIVE_FIELDS = [
    'password',
    'passwd',
    'secret',
    'token',
    'access_token',
    'refresh_token',
    'api_key',
    'apikey',
    'authorization',
    'auth',
    'cookie',
    'session',
    'credit_card',
    'card_number',
    'cvv',
    'ssn',
    'social_security',
]

DEFAULT_SENSITIVE_HEADERS = [
    'authorization',
    'cookie',
    'set-cookie',
    'x-api-key',
    'x-auth-token',
    'x-access-token',
    'x-csrf-token',
]

# Global config
_config = PrivacyConfig()


def configure_privacy(config: PrivacyConfig) -> None:
    """Configure privacy settings."""
    global _config
    _config = config


def get_privacy_config() -> PrivacyConfig:
    """Get current privacy config."""
    return _config


# ===== PII SCRUBBING =====

def scrub_pii(text: str) -> str:
    """Remove PII from text using configured patterns."""
    if not text or not _config.enabled:
        return text
    
    result = text
    for pattern in _config.pii_patterns:
        result = pattern.sub('[REDACTED]', result)
    
    return result


def scrub_dict(data: Dict[str, Any], depth: int = 0, max_depth: int = 10) -> Dict[str, Any]:
    """
    Recursively scrub sensitive fields from a dictionary.
    """
    if depth > max_depth:
        return {"_truncated": True}
    
    if not isinstance(data, dict):
        return data
    
    result = {}
    
    for key, value in data.items():
        lower_key = key.lower()
        
        # Check if field is sensitive
        if any(sensitive in lower_key for sensitive in _config.sensitive_fields):
            result[key] = '[REDACTED]'
        elif isinstance(value, dict):
            result[key] = scrub_dict(value, depth + 1, max_depth)
        elif isinstance(value, list):
            result[key] = [
                scrub_dict(item, depth + 1, max_depth) if isinstance(item, dict) 
                else scrub_pii(str(item)) if isinstance(item, str) 
                else item
                for item in value
            ]
        elif isinstance(value, str):
            result[key] = scrub_pii(value)
        else:
            result[key] = value
    
    return result


def scrub_headers(headers: Dict[str, str]) -> Dict[str, str]:
    """Scrub sensitive headers."""
    result = {}
    
    for key, value in headers.items():
        lower_key = key.lower()
        
        if lower_key in _config.sensitive_headers:
            result[key] = '[REDACTED]'
        else:
            result[key] = scrub_pii(value)
    
    return result


def scrub_url(url: str) -> str:
    """
    Scrub sensitive query parameters from URL.
    Also normalizes IDs to prevent high cardinality.
    """
    from urllib.parse import urlparse, parse_qs, urlencode, urlunparse
    
    try:
        parsed = urlparse(url)
        
        # Scrub query parameters
        query_params = parse_qs(parsed.query)
        scrubbed_params = {}
        
        for key, values in query_params.items():
            lower_key = key.lower()
            if any(sensitive in lower_key for sensitive in _config.sensitive_fields):
                scrubbed_params[key] = ['[REDACTED]']
            else:
                scrubbed_params[key] = [scrub_pii(v) for v in values]
        
        # Normalize path (replace UUIDs and numeric IDs)
        path = parsed.path
        # Replace UUIDs
        path = re.sub(
            r'[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}',
            ':id',
            path,
            flags=re.IGNORECASE
        )
        # Replace numeric IDs
        path = re.sub(r'/\d+', '/:id', path)
        
        return urlunparse((
            parsed.scheme,
            parsed.netloc,
            path,
            parsed.params,
            urlencode(scrubbed_params, doseq=True),
            ''  # No fragment
        ))
    except Exception:
        return scrub_pii(url)


def truncate_payload(payload: str) -> str:
    """Truncate payload to configured max size."""
    if len(payload) <= _config.max_payload_size:
        return payload
    return payload[:_config.max_payload_size] + '...[TRUNCATED]'


# ===== ANONYMIZATION =====

def anonymize_id(user_id: str) -> str:
    """
    Anonymize a user ID for metrics while maintaining consistency.
    Same input always produces same output.
    """
    if not _config.anonymize_user_ids:
        return user_id
    
    # Use HMAC-like construction with salt
    salted = f"{_config.anonymization_salt}:{user_id}"
    hashed = hashlib.sha256(salted.encode()).hexdigest()
    
    # Return first 16 chars for readability
    return f"anon_{hashed[:16]}"


def anonymize_ip(ip: str) -> str:
    """
    Anonymize IP address by zeroing last octet (IPv4) or last 80 bits (IPv6).
    """
    if '.' in ip:
        # IPv4
        parts = ip.split('.')
        if len(parts) == 4:
            return f"{parts[0]}.{parts[1]}.{parts[2]}.0"
    elif ':' in ip:
        # IPv6 - zero last 5 groups
        parts = ip.split(':')
        if len(parts) >= 4:
            return ':'.join(parts[:3]) + ':0:0:0:0:0'
    
    return '[ANONYMIZED_IP]'


# ===== SAMPLING =====

@dataclass
class SamplingRates:
    """Sampling rates for different data types."""
    
    traces: float = 0.1      # 10% of requests
    errors: float = 1.0      # 100% of errors
    slow_requests: float = 1.0  # 100% of slow requests
    metrics: float = 1.0     # 100% of metrics (aggregated anyway)
    
    # Per-endpoint overrides
    endpoint_rates: Dict[str, float] = field(default_factory=dict)


_sampling_rates = SamplingRates()


def configure_sampling(rates: SamplingRates) -> None:
    """Configure sampling rates."""
    global _sampling_rates
    _sampling_rates = rates


def should_sample_trace(endpoint: str, is_error: bool = False, is_slow: bool = False) -> bool:
    """Determine if a trace should be sampled."""
    import random
    
    # Always sample errors if configured
    if is_error and _sampling_rates.errors >= 1.0:
        return True
    if is_error and random.random() < _sampling_rates.errors:
        return True
    
    # Always sample slow requests if configured
    if is_slow and _sampling_rates.slow_requests >= 1.0:
        return True
    if is_slow and random.random() < _sampling_rates.slow_requests:
        return True
    
    # Check endpoint-specific rate
    for pattern, rate in _sampling_rates.endpoint_rates.items():
        if pattern in endpoint:
            return random.random() < rate
    
    # Default rate
    return random.random() < _sampling_rates.traces


# ===== DATA RETENTION =====

class RetentionManager:
    """
    Manages data retention for traces and metrics.
    
    Usage:
        retention = RetentionManager()
        retention.cleanup_expired()
    """
    
    def __init__(self, storage_backend=None):
        self.storage = storage_backend
        self._last_cleanup = datetime.utcnow()
    
    def should_cleanup(self) -> bool:
        """Check if cleanup should run (max once per hour)."""
        return datetime.utcnow() - self._last_cleanup > timedelta(hours=1)
    
    def cleanup_expired_traces(self) -> int:
        """
        Remove traces older than retention period.
        Returns number of traces removed.
        """
        if not self.storage:
            return 0
        
        cutoff = datetime.utcnow() - timedelta(hours=_config.trace_retention_hours)
        # Implementation depends on storage backend
        # Example: self.storage.delete_traces_before(cutoff)
        return 0
    
    def cleanup_expired_metrics(self) -> int:
        """
        Remove metrics older than retention period.
        Returns number of metric points removed.
        """
        if not self.storage:
            return 0
        
        cutoff = datetime.utcnow() - timedelta(hours=_config.metric_retention_hours)
        # Implementation depends on storage backend
        return 0
    
    def cleanup_expired(self) -> Dict[str, int]:
        """Run all cleanup tasks."""
        if not self.should_cleanup():
            return {"skipped": True}
        
        self._last_cleanup = datetime.utcnow()
        
        return {
            "traces_removed": self.cleanup_expired_traces(),
            "metrics_removed": self.cleanup_expired_metrics(),
        }


# ===== LOGGING FILTER =====

class PIIScrubFilter(logging.Filter):
    """
    Logging filter that scrubs PII from log messages.
    
    Usage:
        handler = logging.StreamHandler()
        handler.addFilter(PIIScrubFilter())
    """
    
    def filter(self, record: logging.LogRecord) -> bool:
        # Scrub the message
        if hasattr(record, 'msg') and isinstance(record.msg, str):
            record.msg = scrub_pii(record.msg)
        
        # Scrub args if present
        if record.args:
            if isinstance(record.args, dict):
                record.args = scrub_dict(record.args)
            elif isinstance(record.args, tuple):
                record.args = tuple(
                    scrub_pii(str(arg)) if isinstance(arg, str) else arg
                    for arg in record.args
                )
        
        return True


def configure_pii_logging():
    """Configure all loggers to scrub PII."""
    root_logger = logging.getLogger()
    pii_filter = PIIScrubFilter()
    
    for handler in root_logger.handlers:
        handler.addFilter(pii_filter)
    
    # Also add to specific loggers
    for name in ['uvicorn', 'fastapi', 'sqlalchemy']:
        logger = logging.getLogger(name)
        for handler in logger.handlers:
            handler.addFilter(pii_filter)

File: frontend/lib/performance/privacy.ts
/**
 * Privacy & Sampling Controls for Performance Monitoring
 * 
 * Implements:
 * - PII scrubbing from URLs, payloads, and error messages
 * - Configurable sampling rates per metric type
 * - Data retention policies
 * - User consent integration
 */

export interface PrivacyConfig {
  /** Enable/disable all tracking */
  enabled: boolean;
  
  /** User has consented to performance tracking */
  hasConsent: boolean;
  
  /** Sampling rates by metric type (0-1) */
  sampling: {
    webVitals: number;
    apiLatency: number;
    errors: number;
    userInteractions: number;
  };
  
  /** PII patterns to scrub */
  piiPatterns: RegExp[];
  
  /** URL parameters to redact */
  sensitiveParams: string[];
  
  /** Headers to never log */
  sensitiveHeaders: string[];
  
  /** Max payload size to capture (bytes) */
  maxPayloadSize: number;
  
  /** Data retention period (hours) */
  retentionHours: number;
}

const DEFAULT_CONFIG: PrivacyConfig = {
  enabled: true,
  hasConsent: false,
  
  sampling: {
    webVitals: 0.1,      // 10% of page loads
    apiLatency: 0.05,    // 5% of API calls
    errors: 1.0,         // 100% of errors (important for debugging)
    userInteractions: 0.01, // 1% of interactions
  },
  
  piiPatterns: [
    // Email addresses
    /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g,
    // Phone numbers (various formats)
    /(\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}/g,
    // SSN
    /\b\d{3}[-.\s]?\d{2}[-.\s]?\d{4}\b/g,
    // Credit card numbers
    /\b\d{4}[-.\s]?\d{4}[-.\s]?\d{4}[-.\s]?\d{4}\b/g,
    // JWT tokens
    /eyJ[a-zA-Z0-9_-]*\.eyJ[a-zA-Z0-9_-]*\.[a-zA-Z0-9_-]*/g,
    // Bearer tokens
    /Bearer\s+[a-zA-Z0-9_-]+/gi,
    // API keys (common patterns)
    /(?:api[_-]?key|apikey|access[_-]?token)[=:]\s*["']?[a-zA-Z0-9_-]{20,}["']?/gi,
  ],
  
  sensitiveParams: [
    'token',
    'access_token',
    'refresh_token',
    'api_key',
    'apikey',
    'password',
    'passwd',
    'secret',
    'auth',
    'authorization',
    'session',
    'sessionid',
    'session_id',
    'email',
    'phone',
    'ssn',
    'credit_card',
    'card_number',
  ],
  
  sensitiveHeaders: [
    'authorization',
    'cookie',
    'set-cookie',
    'x-api-key',
    'x-auth-token',
    'x-access-token',
  ],
  
  maxPayloadSize: 1024, // 1KB max
  retentionHours: 24,
};

let _config: PrivacyConfig = { ...DEFAULT_CONFIG };

/**
 * Configure privacy settings
 */
export function configurePrivacy(config: Partial<PrivacyConfig>): void {
  _config = {
    ..._config,
    ...config,
    sampling: { ..._config.sampling, ...config.sampling },
  };
}

/**
 * Get current privacy config
 */
export function getPrivacyConfig(): Readonly<PrivacyConfig> {
  return _config;
}

/**
 * Check if tracking is allowed
 */
export function isTrackingAllowed(): boolean {
  return _config.enabled && _config.hasConsent;
}

/**
 * Set user consent status
 */
export function setConsent(hasConsent: boolean): void {
  _config.hasConsent = hasConsent;
  
  if (typeof window !== 'undefined') {
    try {
      localStorage.setItem('divan_perf_consent', hasConsent ? '1' : '0');
    } catch {
      // Ignore storage errors
    }
  }
}

/**
 * Load consent from storage
 */
export function loadConsent(): boolean {
  if (typeof window === 'undefined') return false;
  
  try {
    const stored = localStorage.getItem('divan_perf_consent');
    if (stored === '1') {
      _config.hasConsent = true;
      return true;
    }
  } catch {
    // Ignore storage errors
  }
  
  return false;
}

// ===== SAMPLING =====

type MetricType = keyof PrivacyConfig['sampling'];

const _sampleDecisions = new Map<string, boolean>();

/**
 * Determine if a metric should be sampled
 */
export function shouldSampleMetric(type: MetricType, key?: string): boolean {
  if (!isTrackingAllowed()) return false;
  
  // Use cached decision for same key within session
  const cacheKey = `${type}:${key || 'default'}`;
  if (_sampleDecisions.has(cacheKey)) {
    return _sampleDecisions.get(cacheKey)!;
  }
  
  const rate = _config.sampling[type];
  const sampled = Math.random() < rate;
  
  // Cache decision
  _sampleDecisions.set(cacheKey, sampled);
  
  return sampled;
}

/**
 * Get effective sample rate for metric type
 */
export function getSampleRate(type: MetricType): number {
  return _config.sampling[type];
}

/**
 * Update sample rate for metric type
 */
export function setSampleRate(type: MetricType, rate: number): void {
  _config.sampling[type] = Math.max(0, Math.min(1, rate));
  // Clear cached decisions for this type
  const keys = Array.from(_sampleDecisions.keys());
  for (const key of keys) {
    if (key.startsWith(`${type}:`)) {
      _sampleDecisions.delete(key);
    }
  }
}

// ===== PII SCRUBBING =====

/**
 * Scrub PII from a string
 */
export function scrubPII(input: string): string {
  if (!input) return input;
  
  let result = input;
  
  for (const pattern of _config.piiPatterns) {
    // Reset regex state
    pattern.lastIndex = 0;
    result = result.replace(pattern, '[REDACTED]');
  }
  
  return result;
}

/**
 * Scrub sensitive parameters from URL
 */
export function scrubUrl(url: string): string {
  try {
    const parsed = new URL(url, 'http://localhost');
    
    // Scrub sensitive query params
    for (const param of _config.sensitiveParams) {
      if (parsed.searchParams.has(param)) {
        parsed.searchParams.set(param, '[REDACTED]');
      }
    }
    
    // Scrub path segments that look like IDs/tokens
    const pathParts = parsed.pathname.split('/');
    const scrubbedPath = pathParts.map((part) => {
      // Keep short segments and common paths
      if (part.length < 20) return part;
      
      // Redact long alphanumeric segments (likely tokens/IDs)
      if (/^[a-zA-Z0-9_-]{20,}$/.test(part)) {
        return '[REDACTED_ID]';
      }
      
      return part;
    });
    
    parsed.pathname = scrubbedPath.join('/');
    
    // Return just path + query for relative URLs
    return parsed.pathname + parsed.search;
  } catch {
    // If parsing fails, do basic scrubbing
    return scrubPII(url);
  }
}

/**
 * Scrub sensitive headers
 */
export function scrubHeaders(
  headers: Record<string, string>
): Record<string, string> {
  const result: Record<string, string> = {};
  
  for (const [key, value] of Object.entries(headers)) {
    const lowerKey = key.toLowerCase();
    
    if (_config.sensitiveHeaders.includes(lowerKey)) {
      result[key] = '[REDACTED]';
    } else {
      result[key] = scrubPII(value);
    }
  }
  
  return result;
}

/**
 * Scrub and truncate payload
 */
export function scrubPayload(payload: unknown): string | null {
  if (payload === null || payload === undefined) return null;
  
  let str: string;
  
  if (typeof payload === 'string') {
    str = payload;
  } else {
    try {
      str = JSON.stringify(payload);
    } catch {
      return '[UNSERIALIZABLE]';
    }
  }
  
  // Truncate if too large
  if (str.length > _config.maxPayloadSize) {
    str = str.slice(0, _config.maxPayloadSize) + '...[TRUNCATED]';
  }
  
  return scrubPII(str);
}

/**
 * Scrub error message and stack trace
 */
export function scrubError(error: Error | string): { message: string; stack?: string } {
  const message = typeof error === 'string' ? error : error.message;
  const stack = typeof error === 'string' ? undefined : error.stack;
  
  return {
    message: scrubPII(message),
    stack: stack ? scrubPII(stack) : undefined,
  };
}

// ===== DATA RETENTION =====

interface StoredMetric {
  timestamp: number;
  data: unknown;
}

const METRICS_STORAGE_KEY = 'divan_perf_metrics';

/**
 * Store metric with retention policy
 */
export function storeMetric(key: string, data: unknown): void {
  if (typeof window === 'undefined') return;
  
  try {
    const stored = localStorage.getItem(METRICS_STORAGE_KEY);
    const metrics: Record<string, StoredMetric[]> = stored ? JSON.parse(stored) : {};
    
    if (!metrics[key]) {
      metrics[key] = [];
    }
    
    metrics[key].push({
      timestamp: Date.now(),
      data,
    });
    
    // Enforce retention policy
    const cutoff = Date.now() - _config.retentionHours * 60 * 60 * 1000;
    metrics[key] = metrics[key].filter((m) => m.timestamp > cutoff);
    
    // Limit stored metrics per key
    if (metrics[key].length > 100) {
      metrics[key] = metrics[key].slice(-100);
    }
    
    localStorage.setItem(METRICS_STORAGE_KEY, JSON.stringify(metrics));
  } catch {
    // Ignore storage errors (quota exceeded, etc.)
  }
}

/**
 * Clear all stored metrics
 */
export function clearStoredMetrics(): void {
  if (typeof window === 'undefined') return;
  
  try {
    localStorage.removeItem(METRICS_STORAGE_KEY);
  } catch {
    // Ignore
  }
}

/**
 * Purge expired metrics
 */
export function purgeExpiredMetrics(): number {
  if (typeof window === 'undefined') return 0;
  
  try {
    const stored = localStorage.getItem(METRICS_STORAGE_KEY);
    if (!stored) return 0;
    
    const metrics: Record<string, StoredMetric[]> = JSON.parse(stored);
    const cutoff = Date.now() - _config.retentionHours * 60 * 60 * 1000;
    
    let purged = 0;
    
    for (const key of Object.keys(metrics)) {
      const before = metrics[key].length;
      metrics[key] = metrics[key].filter((m) => m.timestamp > cutoff);
      purged += before - metrics[key].length;
      
      if (metrics[key].length === 0) {
        delete metrics[key];
      }
    }
    
    localStorage.setItem(METRICS_STORAGE_KEY, JSON.stringify(metrics));
    
    return purged;
  } catch {
    return 0;
  }
}
